Wifi mgmt interface and wifi offload

tbursztyka opened the pull request:
So here is an attempt to expose a WiFi mgmt interface

It tries to solve also how offloaded device will be hooked to the generic WiFi mgmt interface as they do not have any L2.

pnndra committed:
net: Remove the need for an l2 on offloaded drivers

Drivers will be directly contacted via net_if's offload attribute. No
need for a an extra layer as an L2.

Signed-off-by: Dario Pennisi <dario@iptronix.com>
Signed-off-by: Massimiliano Agneni <massimiliano.agneni@iptronix.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfl committed:
net/mgmt: Add initial WiFi management API definitions

Add empty WiFi network management functions that only return -ENETDOWN.
Define management handlers for scan, connect and disconnect requests,
again without any implementation nor parameters defined.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/mgmt/wifi: Add dedicated net mgmt hooks for WiFi offload devices

Exposing connect, disconnect and scan for now.

In case the iface is an instance of a WiFi offload device, the way it
manages scanning, connecting and disconnecting will be specific to that
device (not the mgmt interface obviously). In such case the device will
have to export relevantly a dedicated bunch of function to serve the
mgmt interface in a generic way.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/wifi: Add files skeleton for adding WiFi drivers

This will help not to collide within drivers implementations and/or
avoid dependency from one driver to another one.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@VVESTM have a look. I think we should rebase our respective drivers (your ST disco and my winc1500) on top of that PR.

pfl commented:
So far so good...

VVESTM commented:
@tbursztyka, yes, agree with your proposal. I don't know yet how I will push my PR but on the principle, I agree.
It should avoid merge issue on the driver skeleton...

tbursztyka commented:
I'll try to come up with the parameters today, so we can actually request scan, get the results, connect/disconnect.

GAnthony commented:
Thanks for putting this together.

Since, each vendor's WiFi driver will need to translate its own results into the Zephyr-defined wifi net_mgmt structures,
please take a look at the SimpleLink Wlan API:  https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/ti/simplelink/source/ti/drivers/net/wifi/wlan.h
to help define some of the API parameters and structure fields for wifi net_mgmt.

For example, in SimpleLink, to retrieve wifi scan results:

 ```
SlWlanNetworkEntry_t netEntries[10];
 _i16 resultsCount = sl_WlanGetNetworkList(0,10,&netEntries[0]);
```

where:

```
typedef struct
{
    _u8 Ssid[SL_WLAN_SSID_MAX_LENGTH];
    _u8 Bssid[SL_WLAN_BSSID_LENGTH];
    _u8 SsidLen;
    _i8 Rssi;
    _i16 SecurityInfo;
    _i8 Channel;
    _i8 Reserved[1];
}SlWlanNetworkEntry_t;
```

Also, security parameters:
`SlWlanSecParams_t;`

etc..

I know it will be difficult to handle the variation between vendors, so we'll have to agree on a union of the most relevant data.

GAnthony commented on include/net/wifi_mgmt.h:
```diff
@@ -0,0 +1,83 @@
+
+/* Management part definitions */
+
+#define _NET_WIFI_LAYER	NET_MGMT_LAYER_L2
```
Since an offload driver would bypass L2, I wonder if this definition has any significance?   Is NET_MGMT_LAYER actually used anywhere?

tbursztyka replied:
design wise, the wifi mgmt interface is located in L2. And in the future, if zephyr gets a 802.11 Soft MAC, it will be in L2 as well (the same interface will work there)

tbursztyka commented:
Last patch introduces the parameters for scan results, and connecting. I only added security for open networks and wpa(2) ones. No enterprise, no eap stuff, no wep (one who needs wep will implement the support. But who needs it nowadays?) etc...

No implementation yet.

Only the ssid, channel and rssi (or should it be normalized to something like "signal strength", a value between 0 and 255?).

Maybe such result should be called "network"? and up to the mgmt implementation or driver to gather same bssids/ssid into 1 network (as long as security is the same).

Security type is simpler. No need to tell exactly what cipher or wpa type is behind: psk will do for all of these. While connecting, I see ASF for instance on winc1500 this will work (is does the same global psk ==  wpa(2)), but what about the others?

Tomorrow, I'll try to come up with my winc1500 updated on top of that PR, trying to see how this looks when implemented.

GAnthony commented:
> Security type is simpler. No need to tell exactly what cipher or wpa type is behind: psk will do for all of these. While connecting, I see ASF for instance on winc1500 this will work (is does the same global psk == wpa(2)), but what about the others?

Yes, similarly, for SimpleLink, for all security types, we just have a key and key length pair:

```
typedef struct
{
    _u8   Type;
    _i8*  Key;
    _u8   KeyLen;
}SlWlanSecParams_t;
```

erwango commented:
For Inventek module, scan will report and array of
```
typedef struct {
  WIFI_Ecn_t Ecn;  /*!< Security of Wi-Fi spot. */
  char SSID[WIFI_MAX_SSID_NAME + 1];    /*!< Service Set Identifier value. Wi-Fi spot name */
  int16_t RSSI;                           /*!< Signal strength of Wi-Fi spot */
  uint8_t MAC[6];                   /*!< MAC address of spot */
  uint8_t Channel;                /*!< Wi-Fi channel */
  uint8_t Offset;                  /*!< Frequency offset from base 2.4GHz in kHz */
  uint8_t Calibration;        /*!< Frequency offset calibration */
}WIFI_AP_t;
```

We don't have more info on RSSI, so I guess we'll adapt. Otherwise proposed parameters are fine with us.

Connect will require
```
WIFI_Status_t       WIFI_Connect(
                             const char* SSID, 
                             const char* Password,
                             WIFI_Ecn_t ecn
```);

So seems we're fine with proposal

tbursztyka commented:
I updated the parameters, and provided a quick implementation.

Also, in PR #4711 I added a patch to implement such offloaded API. But had no time to test anything yet.

Anyway, it's just provided as an RFC for now. For instance maybe we want to clarify what could be the status code (see struct wifi_status), currently it's either 0 or -EIO. But maybe we need dedicated codes there. Let's see

tbursztyka commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -18,18 +18,42 @@
 NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_CONNECT, wifi_connect);
 
+static void _scan_result_cb(struct net_if *iface,
+			    struct wifi_scan_result *entry)
```
well, forgot to use this cb

GAnthony reviewed (commented):
A few comments...

tbursztyka commented on include/net/wifi_mgmt.h:
```diff
@@ -0,0 +1,129 @@
+
+enum wifi_security_type {
+	WIFI_SECURITY_TYPE_NONE = 0
+	WIFI_SECURITY_TYPE_PSK,
```
PSK is only for WPA and WPA2.

I came for the most common usage.

WPS, P2P, EAP, ... We will add it later on.

tbursztyka commented on include/net/wifi_mgmt.h:
```diff
@@ -0,0 +1,129 @@
+struct wifi_scan_result {
+	u8_t ssid[WIFI_SSID_MAX_LEN];
+	u8_t ssid_length;
+
```
BSSID is the mac address of the AP. But many AP can serve the same SSID (for roaming for instance). It gets a bit more annoying when security settings are different, then it's a different "network". I am not sure we want to expose that information to the application layer.

Actually, either we group result per-SSID/security, or we expose that bssid and we push all results to the application. Then up to a future ConnMan-like to do the job of presenting networks relevantly.

@pfl  any thoughts on this?

tbursztyka commented on include/net/wifi_mgmt.h:
```diff
@@ -0,0 +1,129 @@
+
+#include <net/net_if.h>
+
+typedef void (*scan_result_cb_t)(struct net_if *iface,
```
We could. But what's the policy of sorting the networks in ext/hal side?  Is it showing only first 10 results it found, is it showing 10 results which have best signal quality, is it callable many times so you can get 10 by 10 results etc... ? I don't see this being easily ported to all ext/hal etc...

VVESTM replied:
You can also check Inventek library in my PR #5075. Interface is in wifi.h file.
For scan, user give a table and a max number : 
    WIFI_Status_t       WIFI_ListAccessPoints(WIFI_APs_t *APs, uint8_t AP_MaxNbr);
Inventek documentation is not really detailed, but in my experience, it returns the most powerfull networks first.

tbursztyka replied:
network as grouped AP serving same SSID/Security, or as for TI: one entry per-AP, even multiple ones serve the same SSID/Security?

GAnthony replied:
For TI ext/hal, first the user sets the scan policy (RSSI threshold, automatic scans and seconds between scans, one-shot scans, etc). Max of 30 results are scanned, which from the examples, appears to be sorted by strength.
Then user can ask for number (<= 30) of results.   
User can call WlanGetNetworkList() repeatedly, to trigger new scans (if one-shot policy was set) and get updated list.
I anticipate that we can have a Kconfig option to set some standard policy (like one-shot), and not expose all those configuration APIs to the user to change at run-time, to try to keep things simple.

jukkar replied:
> Otherwise seems the user must wait for multiple callbacks to build his own scan results table, and will not know when the last callback arrives.

The problem with the pre-allocated list is that it needs memory to be allocated either from stack or statically. We could add a parameter to scan_cb that tells if this is the last result or not.
Would it be possible to have a scan_cb that could have things in both ways, so if the user provides an array, then that is used to populate the result, and if the array is omitted, then callback would be called for each result?

GAnthony replied:
> Would it be possible to have a scan_cb that could have things in both ways
That would be nice, as the offload device code could be simplified, and not be forced to loop through its internal table of scan results calling the scan_cb multiple times.

dleach02 commented:
Is this interface being defined meant to be the highest level interface to a WiFi subsystem? For example, if we were to add a supplicant to the system would this interface be to the suplicant while currently it is likely linked to an offloaded WiFi where the supplicant is in the offload module?

GAnthony commented:
It's my understanding this constitutes the beginnings of the 'control plane' interface.
The general framework documented here:  
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/subsystems/networking/network-management-api.rst
Since the TI offload chip contains the supplicant, known enterprise certificates are burned into secure flash ahead of time with a known filename, so no need to pass (eg: PEM) filenames or store keys as connect parameters.  For a non-offloaded supplicant, not sure how that will look.

dleach02 commented:
The reason I ask is that the classical IEEE802.11 layering there is the MLME-PLME SAP to the MAC which the SME uses to control the MAC and then you have the SAP interface to the SME. In the Linux world you end up with the interface to the WPA supplicant and the close approximation to the MLME-PLME SAP would be the NL80211 interface or cfg80211_ops. The WPA supplicant SAP interface would be the next layer up for control/management of 802.11. So looking at those interfaces as reference would be useful in considering what we want to put into Zephyr (or roadmap). Now I'll admit that the first (likely only) 802.11 enablement in Zephyr will be from these connected modules so I would suggest that the layer we are trying to define would be the SME SAP layer.

jukkar commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -18,6 +18,13 @@
 			void *data, size_t len)
 {
+	if (net_if_is_ip_offloaded(iface)) {
+		struct net_wifi_mgmt_offload *off_api =
```
The "off" prefix looks unnecessary here, why not just call it "api"?

jukkar commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -27,6 +34,13 @@ NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_CONNECT, wifi_connect);
 		     void *data, size_t len)
 {
+	if (net_if_is_ip_offloaded(iface)) {
+		struct net_wifi_mgmt_offload *off_api =
```
ditto

jukkar commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -36,6 +50,13 @@ NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_SCAN, wifi_scan);
 			   void *data, size_t len)
 {
+	if (net_if_is_ip_offloaded(iface)) {
+		struct net_wifi_mgmt_offload *off_api =
```
and here too.

jukkar reviewed (commented):
Looks very good, couple of comments.

pfalcon commented:
@tbursztyka, @GAnthony, @erwango : Any progress on this? Would be nice to get (something) of this into 1.11, with feature freeze in a week. If there're contention points, I'd suggest to define a minimal scope of what might fit, and ask @tbursztyka to prune initial version for this release. Otherwise, let's get approvals up? ;-)

@tbursztyka : Can you please rebase and fix conflict?

GAnthony commented:
@pfalcon: I believe @tbursztyka mentioned in networking forum that this PR is waiting for test hardware to validate.  I offered to try it out on cc3220sf, but probably won't get to it until next week.  Until then, I think it may be premature to approve without some testing.

pfalcon commented:
@GAnthony: Oops, I missed that, audio wasn't good for me. What I got is that it's a week before the feature freeze, and I see that @carlescufi , as the release manager for 1.11, already started triaging of 1.11-milestoned tickets. I also wanted to see what's in the queue, and thought that a friendly ping on this PR might help...

GAnthony commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -0,0 +1,109 @@
+}
+
+static int wifi_scan(u32_t mgmt_request, struct net_if *iface,
+		     void *data, size_t len)
```
Maybe specifically pass the `wifi_scan_result` array and size here:

```
static int wifi_scan(u32_t mgmt_request, struct net_if *iface,
		     struct wifi_scan_result *entries, size_t len)
```

GAnthony commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -0,0 +1,109 @@
+		return;
+	}
+
+	net_mgmt_event_notify_with_info(NET_EVENT_WIFI_SCAN_RESULT, iface,
```
@@In the case of SimpleLink wifi scan, the table of scan results will be available from the chip all at once.   So, during the wifi_scan() call, the _scan_result_cb() can be called up to 30 times (max) in rapid succession.

Now, the net_mgmt server thread uses an event queue to store notifications for later delivery.

Depending on the values of 
`CONFIG_NET_MGMT_EVENT_QUEUE_SIZE (default 2)`,
and `CONFIG_NET_MGMT_EVENT_THREAD_PRIO (default 7)`,
and the priority of the thread calling wifi_scan(), 
it seems the server thread could drop notifications.

One solution would be to force `CONFIG_NET_MGMT_EVENT_QUEUE_SIZE` to be the maximum table size (30) in this case; but this will be wasting extra memory - especially for an operation that is infrequently executed.

So, @jukkar 's suggestion of passing in the scan_results table would nicely avoid that issue here...

GAnthony reviewed (changes requested):
FYI: I started (re)implementing a wifi offload driver for TI SimpleLink based on this PR.   Some more comments on the wifi_scan() operation...

Commit:
net: wifi: Add a wifi.h to hold WiFi definitions

Placeholder to hold generic WiFi definitions.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
net: Don't call l2->enable for offload devices

Previously, with CONFIG_OFFLOAD enabled, net_if_up() and
net_if_down() were calling l2->enable, which didn't exist,
so was crashing.
Instead, if CONFIG_OFFLOAD is enabled, it will test for if the net_if is
of an offload type and react accordingly.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/wifi: Add a shell module for controlling WiFi devices

This module exposes for now the net mgmt for WiFi: connect, disconnect
and scan commands.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/wifi: Select dependencies to get wifi mgmt working

Wifi mgmt interface will be required by offloaded wifi device drivers
therefore when selecting this interface, make sure dependencies are
selected too.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/mgmt: Move NET_EVENT_INFO_MAX_SIZE into net core's private header

First because nobody needs to know that besides net_mgmt core and
secondary to avoid possible circular dependancy on
net_mgmt.h/net_event.h.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@GAnthony, @erwango :  I rebased this PR according to @GAnthony 's PR #6466 findings, though I also fixed things differently (to know when scan is over, so no more results to come: I added an event).

I also added a wifi shell module so it's simpler to test. See PR #4711  I added a sample for it, it could get expanded to support SimpleLink as well etc...

GAnthony commented on subsys/net/ip/l2/wifi_shell.c:
```diff
@@ -0,0 +1,207 @@
+	const struct wifi_scan_result *entry =
+		(const struct wifi_scan_result *)cb->info;
+
+	printk("SSID: %s (%u) - Channel %u - Security: %s - RSSI: %d\n",
```
Would it be possible to print out a nicely formatted table, as done in this commit: https://github.com/zephyrproject-rtos/zephyr/pull/6466/commits/667d228204470cb006c9fd6dd167cdd698b16a84  ?

GAnthony commented on subsys/net/ip/l2/wifi_shell.c:
```diff
@@ -0,0 +1,207 @@
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		printk("Disconnection request failed (%d)\n", status->status);
```
While testing I found that a disconnect event can occur without a disconnect request, for example if the AP disconnects the device via an error or removing a hotspot on a mobile phone.
I found it can also occur during a connect after a previous connect to the same SSID (i.e., the AP was disconnecting, then starting the reconnect).
So, perhaps we just print here "Disconnected".

tbursztyka replied:
Yep, forgot that obvious use case when the AP dies, or becomes too far etc etc...

GAnthony reviewed (commented):
Thanks for writing this.   (I attempted a net_wifi shell, but have usage faults on my board with net_shell, so couldn't push up).
Just a few minor comments...

GAnthony commented:
One comment I had from a TI colleague:  Could we also have connect/disconnect events for ethernet? 
In which case, maybe that could be common to net_mgmt.h?

tbursztyka commented:
@GAnthony ethernet connect/disconnect, you mean carrier detection? Sure, but that's not for wifi mgmt, see issue #6640

GAnthony commented:
If carrier detection includes notification when the cable is connected/disconnected, then, yes I guess that would cover it.

tbursztyka commented:
Ok will add this requirement to the issue.

GAnthony commented on subsys/net/ip/l2/wifi_mgmt.c:
```diff
@@ -0,0 +1,125 @@
+	     (params->channel > WIFI_CHANNEL_MAX)) ||
+	    !params->ssid || !params->psk) {
+		return -EINVAL;
+	}
```
@tbursztyka , this code failed when I tested by attempting to connect to an SSID with open security.

Should the intended test should be something like this?

```
	if (((params->security >= WIFI_SECURITY_TYPE_PSK) &&
	     (!params->psk || (params->psk_length < 8) || 
	      (params->psk_length > 64))) ||
	    (params->ssid_length > WIFI_SSID_MAX_LEN) ||
	    ((params->channel != WIFI_CHANNEL_ANY) &&
	     (params->channel > WIFI_CHANNEL_MAX)) ||
	    !params->ssid) {
		return -EINVAL;
	}
```

tbursztyka replied:
yep. I haven't been able to test this yet.
In fact I am still struggling to get winc1500 to connect to anything, this crap always fails authenticating and does not tell why. (formerly connecting with static credentials used to work)

GAnthony replied:
https://store.ti.com/CC3220SF-LAUNCHXL-SimpleLink-Wi-Fi-CC3220SF-Wireless-Microcontroller-LaunchPad-Development-Kit-P50638.aspx
:-)

tbursztyka replied:
:D

GAnthony commented on subsys/net/ip/l2/wifi_shell.c:
```diff
@@ -0,0 +1,244 @@
+		return -EINVAL;
+	}
+
+	cnx_params.ssid = &argv[1][1];
```
This change keeps the first letter of the SSID from being stripped:
```
-       cnx_params.ssid = &argv[1][1];
+       cnx_params.ssid = argv[1];
```

tbursztyka replied:
Hum yes that's done on purpose : ssid should be given starting with " and ending with ". see help  of the wifi shell module. It's because ssid can hold spaces.

tbursztyka replied:
In an ideal implementation of the wifi shell, it would parse the SSID as a byte array. I did not bother doing that yet since neither simplelink nor asf treat ssid as a byte array as _they should_ because the wifi spec says so. But if we get a native wifi stack someday, it will be necessary.
As we don't store the results anywhere, it's not possible to use another information (such as a hash, or an index etc...). So length is needed.

GAnthony replied:
> I did not bother doing that yet since neither simplelink nor asf treat ssid as a byte array as they should because the wifi spec says so. 

OK.
Actually, SimpleLink connect API *does* treat the SSID as a byte array, and of a given length.
``_i16 sl_WlanConnect 	( 	const _i8 *  	pName,
		const _i16  	NameLen,
		const _u8 *  	pMacAddr,
		const SlWlanSecParams_t *  	pSecParams,
		const SlWlanSecParamsExt_t *  	pSecExtParams 
	) 	
``

GAnthony replied:
@tbursztyka, FYI, just got around to testing wifi_shell connecting to an AP with a space embedded in the SSID name.
Turns out, shell module line2arg() function doesn't support quotes.
`wifi> connect "HVL Network" 11 0 password`
yields tokens:
`{"HVL, Network", 11, 0, password}`
Filed https://github.com/zephyrproject-rtos/zephyr/issues/7178 , maybe the shell owner can add as an enhancement.

GAnthony reviewed (changes requested):
Some issues I encountered trying to integrate SimpleLink with this interface, and some recommendations.  

Issue 1: The offload chip does not return wifi_mgmt scan results asynchronously:

SimpleLink by default periodically scans, keeping an internal table updated.   One can disable the auto scan, making it work in "one-shot" mode (which is what I do).  But in either case, the client gets the results table via a synchronous call to sl_WlanGetNetworkList().
There is *no* asynchronous event notifying the MCU that the scan is completed.
Therefore, to fit this wifi_mgmt scan API, such chips would need to create an extra server thread or workqueue to poll for the results to send them asynchronously, or just return the results during the scan request (as I do, but then this leads to Issues 2 & 3).  

Recommendation:  Simplify the logic to:
     a) scan_start API:          start the scan.
     b) get_scan_results API: return table of results, or -EAGAIN if results not ready.

Issue 2: Offload chip returns table of results, not one result at a time:

This requires an extra loop in each wifi driver to trigger each NET_EVENT_WIFI_SCAN_RESULT notification, followed by code to trigger the NET_EVENT_WIFI_SCAN_DONE notification.
Recommendation:
     - get_scan_results API.
From past comments, I recall the STM WiFi chip worked similarly for the scan, so a more direct API might help there as well.

Issue 3: net_mgmt notifications may get dropped.

If there are more events than the size of CONFIG_NET_MGMT_EVENT_QUEUE_SIZE, events can get dropped.  Increasing that config variable is not obvious: it needs to be the max of all subsystems' needs using the net_mgmt API, even then it's still possible to loose events under some scenarios.  
Some events may be important to decide a next action (like a disconnect event
enabling a socket operation to return an error), so events must not be dropped.
We can require each WiFi driver to force each notification to be processed by explicitly calling k_yield() after each notification to run the net_mgmt callbacks, but that seems like a workaround.

Recommendations:
 - Re-architect net_mgmt to get rid of the net_mgmt thread, or make into a workqueue, or maybe force k_yield() to run the net_mgmt when the event queue gets full?
  - Or at least, raise an assert when an event gets dropped, as this is an error condition.

tbursztyka commented:
> Issue 1: The offload chip does not return wifi_mgmt scan results asynchronously:

> SimpleLink by default periodically scans, keeping an internal table updated. One can disable the auto scan, making it work in "one-shot" mode (which is what I do). But in either case, the client gets the results table via a synchronous call to sl_WlanGetNetworkList().
There is no asynchronous event notifying the MCU that the scan is completed.

I would blame simplelink here, that's bad design.

> Therefore, to fit this wifi_mgmt scan API, such chips would need to create an extra server thread or workqueue to poll for the results to send them asynchronously, or just return the results during the scan request (as I do, but then this leads to Issues 2 & 3).

> Recommendation: Simplify the logic to:
>a) scan_start API: start the scan.
>b) get_scan_results API: return table of results, or -EAGAIN if results not ready.

That's what happens when a framework is designed to be used directly and not below another API. :(

I am not going to introduce such get_scan_results thingy: too heavy. It means that targeted devices has to reserve enough space for handling these, in addition of the space reserved by the offloaded driver. That's a no-go.

Scanning is by definition an asynchronous task.

> Issue 2: Offload chip returns table of results, not one result at a time:

> This requires an extra loop in each wifi driver to trigger each NET_EVENT_WIFI_SCAN_RESULT notification, followed by code to trigger the NET_EVENT_WIFI_SCAN_DONE notification.
Recommendation:
>- get_scan_results API.
>From past comments, I recall the STM WiFi chip worked similarly for the scan, so a more direct API might help there as well.

Stick with the loop.

> Issue 3: net_mgmt notifications may get dropped.

> If there are more events than the size of CONFIG_NET_MGMT_EVENT_QUEUE_SIZE, events can get dropped. Increasing that config variable is not obvious: it needs to be the max of all subsystems' needs using the net_mgmt API, even then it's still possible to loose events under some scenarios.
Some events may be important to decide a next action (like a disconnect event
enabling a socket operation to return an error), so events must not be dropped.
We can require each WiFi driver to force each notification to be processed by explicitly calling k_yield() after each notification to run the net_mgmt callbacks, but that seems like a workaround.

>Recommendations:

 >   Re-architect net_mgmt to get rid of the net_mgmt thread, or make into a workqueue, or maybe force k_yield() to run the net_mgmt when the event queue gets full?
    Or at least, raise an assert when an event gets dropped, as this is an error condition.

net_mgmt is a bit like netlink in Linux: there is no guaranty you will get what you are waiting for. It's just not meant for that. (Btw workqueue is not going to solve anything here).
Using k_yield() is up to the net_mgmt user, and I actually advise to do so for scanning results. I used it in winc1500. (Your thread going through the scan results loop would call k_yield() after each event thrown).
net_mgmt is not going to do the k_yield() by itself for the good reason that it can't tell if the caller is on a critical path or not, so it should not decide by itself if it's ok to pause the current task.

CONFIG_NET_MGMT_EVENT_QUEUE_SIZE is a critical option yes, but we can't come with a good enough default for everyone, it's just not possible as there are so many use cases around. At least I don't see how to solve that nicely. It could be better documented though.


Note that when @pfl  and myself designed wifi mgmt we did not do it with a particular offload API in mind (I did not even knew how atmel's ASF work around that), but we based it on our own experience with Linux and WiFi (quite some years of it).

Such interface must not be designed bottom to top anyway.

tbursztyka commented:
@GAnthony Could you unlock the PR? unless you see obvious bugs to fix. Let's apply this PR, we'll fix if anything comes later.
